### ISSUE TITLE
Some File Format Notes/Choices

### ISSUE THREAD

**Issue by jmdelahanty** [Tyelab/bruker_pipeline#24]
Since I haven't documented what I've learned about different containers/file formats for data (ie H5/zarr) in one place yet I figured I'd just plop down things here. Note that this is just what I've learned since joining the lab from docs/asking people questions, so there could be inaccuracies/misunderstandings.

# HDF5: Hierarchical Data Format 5

The [HDF group](https://www.hdfgroup.org/) created the hierarchical data format in 1985 to handle large datasets for supercomputing. It's been in wide use for many decades as a lossless container of multiple types of data streams. Each of these data streams can be interacted with using the spec's integrations with MATLAB, Python, C++, etc. The library has been maintained and standardized over all this time to make it easy for use and adoption across a wide variety of different use cases. The format is a flat binary representation of whatever is inside the file with metadata that maps to where in the bit stream different things are.

## HDF5 is Lossless, Unless You Tell It Not To Be

One thing that was mentioned in our recent meeting was that HDF5 is always lossy unless you configure it specifically not to be. This is not in line what I've seen/read in my experience so far, so I wanted to write down what I've learned about this part here as means of having documented discussions about it.

Here's what [John Kirkham](https://github.com/jakirkham), a software engineer at NVIDIA and former Janelia Research Campus engineer, told me:

> Meaning HDF5? I don't think so. Believe it is lossless. Don't think it even compresses by default. One would need to opt-in.

Here's what [Talmo Periera](https://github.com/talmo), SLEAP lead developer who uses H5 formats extensively and a professor here at Salk, told me:

> HDF5 is 100% lossless

Here's what [Pat Gunn](https://github.com/pgunn), a lead developer/maintainer of the [CaImAn](https://github.com/flatironinstitute/CaImAn) library, told me:

> I was not aware of any data types in hdf5 that are lossy. I'd ordinarily assume that any data stored in hdf5 is done so in a lossless way, or someone has a truly weird scheme requiring outside interpretation to do lossy reads of what's stored in the file. (although hdf5 is as complex as tiff, sadly, so maybe there are some lossy variants)

Each of these experts had told me to check out [this](https://docs.h5py.org/en/stable/high/dataset.html#lossless-compression-filters) link for more information about the container's compression.

If one wants to enable a form of lossless compression like `blosc` or `lzf`, you perform [chunking](https://confluence.hdfgroup.org/display/HDF5/Chunking+in+HDF5) so the compression operation can be performed in parallel upon chunks of data instead of serially.

You can of course apply a lossy compression algorithm if you'd like, but from what I've seen so far you need to specify a compressor that does this.

In my experience so far, if you just tell a library like `h5py` to create an HDF5 store, it doesn't perform compression of any sort. If you compare the size of input tiffs to a resulting H5 store using default settings you end up with datasets that are the same size!

## HDF5 Allows for Parallel Reads Easily, Parallel Writes Less So

A nice advantage of the HDF5 format is that it allows for parallel reads and, with some extra configurations/compiling the library in different ways, you can also perform parallel writes to the datasets.

A nice advantage of parallel reads from the same object is that multiple programs/threads/processes can access the same data nearly at the same time in a way that handles concurrency without us having to write methods to do it. When we're implementing these cool genetic programs for finding the best algorithms to use for our data, being able to perform parallel reads so we can implement the algorithms in parallel could help make things more efficient than if we were to do them in series. This can be especially helpful for things that would not fit entirely into memory on our machines.

The data we're working with wouldn't be large enough most likely to cause memory problems on our powerful computers, but I think we could perform many different kinds of work all at once by doing parallel reads from the object instead of loading the data entirely into memory and then doing the algorithmic work upon it. This is the sort of thing that Dask does via Dask Arrays I think...

## HDF5 Allows for Quick Integration with [Dask](https://docs.dask.org/en/stable/index.html) For Parallel Computing

The Dask library is immensely powerful and allows for the execution of complex data processing in parallel across many machines. There are simple methods that Dask has implemented to go from H5 files into Dask arrays that allow for high performance computation on large files. See [here](https://docs.dask.org/en/stable/generated/dask.dataframe.read_hdf.html#dask-dataframe-read-hdf).

# Zarr

The [zarr](https://zarr.readthedocs.io/en/stable/index.html) library was inspired by the N5 and HDF5 formats and designed to be a Python centric implementation of a data storage format that allows for parallel reads and writes by design for very large datasets. Many projects that are performing computations on huge/larger than memory datasets (such as connectomic electron microscopy and spatial transciptomics data) rely on zarr for accelerating their computation. It has superb cloud support as well and has been designed for use with cloud storage in mind. Further, it is also easily integrated into parallel computing libraries like Dask.

## Zarr is Lossless, Unless You Tell It Not To Be. It's also really fast.

The same information that's in the HDF5 section above is applicable here for zarr. One design choice of zarr is that instead of storing things as a long flat binary, it instead stores things as small chunks in a nested directory structure. This allows for parallelism upon chunks that is *extremely* fast when you're performing computation.

For example, generating a compressed HDF5 dataset out of one of Austin's recordings' 45k tiffs using lossless blosc compression took about 4.5 minutes on a beefy machine in the datacenter (tako). Using zarr, it took just over 1 minute! See [here](https://github.com/Tyelab/bruker_pipeline/issues/9#issuecomment-996304859) for some notes.

## Zarr Allows for Parallel Reads *and* Writes

A huge benefit of zarr is that it allows for both parallel reads and writes out of the box. This means that we could organize a dask/[daisy](https://github.com/funkelab/daisy) scheduler that's performing work on our chunked datasets in parallel and alter the data as it's being worked on continuously. When we're implementing the tools that Annie is developing, it might be really nice to have a dataset that can have parallel writes done to it so we can run algorithmic testing in parallel.

Suite2p has labeled the use of zarr as [an enhancement](https://github.com/MouseLand/suite2p/issues/795) that can maybe speed up their already speedy algorithm.

## Zarr Allows for Quick Integration with Dask For Parallel Computing

Just as you can use Dask to load H5 files into Dask Arrays, you can use Dask to map zarrs to Dask Arrays too! See [here](https://docs.dask.org/en/stable/generated/dask.array.from_zarr.html) for the docs.

## A Disadvantage of Chunked Zarr Datasets

Chunked datasets can end up being difficult to move around on networks because typically chunks are quite small stored as a nested directory structure as small files. Here's what [Arlo Sheridan](https://github.com/sheridana), a software engineer and deep learning researcher who introduced me to zarr, told me:

> The only annoying thing with zarr is storage and sending them around because of the nested directory structure. It makes sense for single copies of large data that stay confined to some location. But if you are constantly generating tons of datasets that all fit in memory then it is probably overkill. [A 25GB timeseries (a size slightly larger than the data we collect for Austin's project at least)] would make sense storign as a zarr and doing chunk writes too with Dask or daisy. But if you are going to constantly be generating 25GB datasets like every day then maybe not.

Arlo also pointed me to [this](https://openorganelle.janelia.org/tutorials#python) project from Janelia which collects huge amounts of data does visualization/processing/data access via Python and zarrs or N5 as an example. These datasets are far far larger than what we collect/use, but could be interesting to check out.

However, much of the automated pipeline that is being designed is likely going to be done on the local scratch space on the compute nodes which have SSD/NVMe storage on board. If things aren't entirely stored on local scratch spaces, we would only incur a small network transfer penalty when the program starts, so at the very beginning, and not again. This would be especially true if we used Dask since the Dask Array stores that results are placed into would be on the compute node's local scratch space.

# Using H264/H265 Will Require Modifications, Might Lose Data (Data Lost Might Be Unneeded!)

The H264/5 encoders that we will be adopting have behaviors that should probably be considered by the lab given the properties of our data and how the data is acquired.

## H264/H265 Does *Not* Encode 16bit Data Natively. We Also Don't Use 16bits/8bit may be fine!

One of the properties of how the microscope outputs data is that each pixel has a possible value between 0 and 65535, or 16 bits. It should be noted that H264/H265 is not capable of encoding 16bit datastreams directly. There are ways of multiplexing data across channels (ie 8 bits stored in a red channel, the other 8 in a green channel), but doing so may make interacting with the data more difficult for people not familiar with how to access things in that manner. From what others have taught me/looking around online, the H265 codec can encode up to 12bit values, or 0 - 4096.

While the data could inhabit the range between 0 and 65535 per pixel, images never reach the ceiling. Something would probably be wrong if we saw that in our data. I'm not currently aware of what range of intensities are present throughout an entire recording.

Here's some example images from a high signal to noise recording with calibration bars illustrating the values of pixel intensities:

16bit, one tiff

![single_tiff_16bit](https://user-images.githubusercontent.com/28411193/174686771-ff453607-322a-46ee-bfe3-a47fc9e61cf2.png)

You can see here that the range of values for this example tiff goes from 0 through 6143, which is beyond the 12bit capabilities of H265.

An average of these tiffs shows a slight decrease in ranges of the image, from 0 to 4273, just beyond the 12bit capabilities of H265 encoding.

![average_tiff_16bit](https://user-images.githubusercontent.com/28411193/174686974-9a3df589-d3ed-418d-b8b5-66e9e38faa32.png)

Here's what an example tiff rescaled to 8 bits looks like:

![single_tiff_8bit](https://user-images.githubusercontent.com/28411193/174686966-8a14c588-e6e2-407e-a64e-54e42e08517f.png)

It looks quite similar, but we would probably want to empirically test whether or not having the range of available pixel values going from it's original representation to a different bit depth would alter the conclusions we get out of the data. I know that Chris for example takes 16bit encoded images from Inscopix and scales them to 8bit without altering analysis in a harmful way, but I don't know if doing that in 2P imaging will behave the same. Here's what he said about this:

> Rationale is that you probably don’t need the 2^16 dynamic range to capture bursts of action potentials

This makes some sense to me and would definitely make datasets much smaller/faster to do work on.

Here's something that Jonny Saunders, a graduate student at the University of Oregon who has been an immensely helpful resource for learning about all these complicated encoders (also the author of Autopilot!), has told me about what we're considering:

> There are multiple ways of representing 16-bit samples: eg. there is a 16-bit grayscale pixel format (gray16be and gray16le for big-endian and little-endian) but it's not supported by x264 or x265, not sure if any video codecs do. You can pack the 16 bits in two other channels (eg. red and green are both stored as 8 bits, so put the first 8 bits in the red and the second in green) but that can do funky things with the psychophysical model of most compression algorithms. the entire design of compression algorithms, like why they can compress the data more than just gzipping the frames, is that they have a psychophysical model that discards data that the visual system normally doesnt perceive. Most of the time this doesn't matter, but for microscopy data/quant data you want to be a bit careful and definitely do it in a test-driven way... People are sort of right to be thoughtful about encoding and compression because as above the algorithms have specific psychophysical assumptions/etc. It's not really about the type of camera as much as the type of video: quantifiable image series are different than movies and have different concerns for storage, but it's an empirical question what the best balance of compression, I/O performance, etc.

I also reached out Dr. Andre Andreev, a post-doctoral fellow at Cal Tech who has been teaching me different microscopy things for a while, and he illustrated a nice little figure that can help show the idea of rescaling/trimming bit depths from data [here](https://twitter.com/messages/2909395381-942394886673727488).

This brings up questions about what it was that H264/H265 video encoders were designed for: the `psychophysical model` of the algorithm - an idea I had no idea existed for this sort of thing.

## H264/H265 Encoders Might Not Be Appropriate For Encoding This Data?

Throughout my asking for advice/opinions on the use of these encoders for our data, I've tried being very explicit that we would be losslessly encoding things. However, each person I've asked has brought up that these video codecs are not commonly used for lossless encoding of data because the codecs were designed for the explicit purpose of being lossy compressors.

I'm guessing that the way I communicated about these things wasn't clear, because without fail people told me that we should take extreme caution when using them for our microscopy data even when I say that we want to do it without any compression being done.

Here's the different things people have told me about these codecs for microscopy data.

Since Talmo knows these video codecs extremely well, I reached out to him to see what he thought about encoding things this way. He advised strong caution when using these encoders for the following reasons (paraphrasing):

- Lossless H264 doesn't seem to play nice with the codec
- You would have to quantize the data to 8bit or multiplex through the color channels, which he said can introduce different behavior of analysis software
- H264 was designed to compress motion, not fluorescence point scans captured via photodetectors. While these codecs might have artifacts that are imperceptible to the eye, it can break downstream processing like roi detection.

Professor Jeff Hardin from the University of Wisconsin had this to say. Note that I clarified in a response to this that we were going to use lossless encoding, but he brought up the lossy nature of their typical use:

> The reasons people have given you for avoiding H.264/5 are right if you’re dealing with confocal/multiphoton/lightsheet data. You will lose information. If you know what information you don’t need and can throw away, however, this can be OK. My former colleague here at Wisconsin, Jan Huisken, has made this move for the huge lightsheet datasets his lab generates. If you know you only needs certain aspects of the data, then the savings when it comes to archiving these datasets can be enormous. However, there’s no going back once data is destroyed and discarded. 
Ffmpeg and various codecs like H.264/5 are also useful if you want to compress data for playback of movies in something like PowerPoint or supplemental material for a paper. Then mildly lossy compression is fine, since the goal is illustration, not quantification. 
I’m not sure what sort of 2P data you’re generating, so my advice is that if you know that compression won’t compromise the data you want to analyze later, H.264/5 could be OK, but extreme caution is warranted if you plan to do post-acquisition quantification/analysis of fluorescent signals.

I also messaged developer of the OME-NGFF format, [Josh Moore](https://github.com/joshmoore), about this sort of thing and here's what he had to say:

> Depending on your use case, H264 might work for you. but it's a very different toolset.

Here is what Sammy Wiser-Novak from our Biophotonics Imaging Core had to say. (Note again that he discussed the lossy use of the codec, so I provided additional clarification about it and will update this).

> I believe that the codecs you are suggesting to use will compress your data according to their described standards.

In the end, the common thread that everyone has given me so far re what to do is essentially try it out and see what happens. If it looks okay/behaves normally, then we're good to go! But we should probably test it out first.

## There Are Ways of Writing Out `.bin` without H264/5: Suite2p

Suite2p outputs a `.bin` file and performs analysis on it using [this](https://github.com/MouseLand/suite2p/blob/main/suite2p/io/binary.py) code. For some reason they don't use these video encoders. I've emailed Dr. Stringer about it but haven't heard back yet/might not since she's so busy... Note also that there's ways of decoding tiffs to binary using the `tifffile` package. See [here[(https://github.com/Tyelab/bruker_pipeline/issues/14#issuecomment-1156877952). 

Note that Suite2p has no method for ingesting data encoded via H264/265 or for zarr, so for us to validate/test things we'll have to make a fork and try out their methods on our data by altering their code I think.

# Things to Do or Test

Given all this, there's a couple things that we might want to test out:

- [ ] Does rescaling data from uint16 to uint8 alter the results of analysis?
- [ ] Does the use of these encoders, regardless of compression or not being enabled, alter the data in some manner that's unexpected? I can't see why it would, but I haven't seen it tested as I've seen zarr and HDF5 tested...
- [ ] How long does it take to write this data to disk from tiffs? Note again that H5 via dask can take 4-5 minutes, zarr 1-2 minutes.
- [ ] Is accessing a bitstream from data encoded this way faster/easier than doing so via H5/zarr?
- [ ] Other things? There's lots of other stuff to consider, so this is by no means a complete list of considerations of course...

**Comment by jmdelahanty** [comment]
When converting an original recording's tiffs into uint8, it looks suite2p analysis works just as well which is cool! I have no idea if the downstream results look the same, but clicking around different ROIs things look comparable/the same to me at least visually. Whether or not the actual results in the end are the same I'm not sure...

Here's some outputs after testing for a little bit today.

# Mean Images

`uint16`

![original_uint16_mean](https://user-images.githubusercontent.com/28411193/174721849-a7b6c8f1-dcb2-4bde-a219-bec3212d293b.png)

`uint8`

![correct_uint8_mean](https://user-images.githubusercontent.com/28411193/174721866-5ec4e90b-f64c-47c8-95e4-fa9c6cebc4e4.png)


# Mean Enhanced

`uint16`

![uint16_mean_enhanced](https://user-images.githubusercontent.com/28411193/174721945-3e070106-5e9f-447a-a3b7-a7cff352e5c9.png)

`uint8`

![enhanced_correct_uint8](https://user-images.githubusercontent.com/28411193/174721992-37667aad-85ef-42f3-a571-41c02b9e07a7.png)

# Correlation Maps

`uint16`

![uint16_correlation_map](https://user-images.githubusercontent.com/28411193/174722088-1b90db67-1703-4c65-a379-6f3d4c3e69fb.png)

`uint8`

![correct_correlation_map_uint8](https://user-images.githubusercontent.com/28411193/174722142-cfbac763-9cc0-479a-86b9-de6842c2c465.png)

# Max Projection

`uint16`

![uint16_max_projection](https://user-images.githubusercontent.com/28411193/174722179-27888215-5eb6-49c9-8633-7e2274499917.png)

`uint8`

![correct_max_projection_uint8](https://user-images.githubusercontent.com/28411193/174722229-255e958a-f4fe-47cb-9aec-4e68c7661057.png)

**Comment by jmdelahanty** [comment]
Here's what things can look like if you don't rescale your data properly. This is using Dask, h5py, and skimage. I found all this out by failing to use these libraries properly for the test above.

Here's what a 16bit image from a good SNR recording looks like for reference: 
![single_tiff_16bit](https://user-images.githubusercontent.com/28411193/174905126-ccc2a68f-43ef-4a8c-bf55-c6f3f3278caf.png)

Here's what the mean image looks like from this dataset, with ROIs over it:
![original_uint16_mean](https://user-images.githubusercontent.com/28411193/174907212-16ce2ff1-8fdb-49d4-9a35-e31033bbd5d0.png)

# Coercing uint16 to uint8 (truncating data)

```
import h5py
import dask.array
import tifffile
from imagecodecs import tiff_decode
from pathlib import Path


def tiff2hdf5(
    hdf5file, tifffiles, dataset_name='2p', chunksize=128, compression=None
):
    """Write images from TIFF files to chunked dataset in HDF5 file."""

    def imread(filename):
        # return first image in TIFF file as numpy array
        with open(filename, 'rb') as fh:
            data = fh.read()
        return tiff_decode(data)

    image_paths = [p for p in Path(tifffiles).glob("*Ch2*.tif")]
    sorted_paths = sorted(image_paths)

    store = tifffile.imread(image_paths, imread=imread, aszarr=True)
    images = dask.array.from_zarr(store)

    images.to_hdf5('/path/to/file.hdf5', '/data', dtype='uint8')
```

If you simply tell Dask to coerce the datatype from 'uint8' to 'uint16', I think it just crops the data to values between 0 and 255. Here's what that ends up looking like as an average image:
![truncated_uint8_mean](https://user-images.githubusercontent.com/28411193/174905701-9750f3e5-1fa7-416a-ac1c-895587b7bed6.png)

As an average enhanced image:
![truncated_uint8_max_projection](https://user-images.githubusercontent.com/28411193/174905768-e107e5f6-b340-4af7-b928-128c486d81e4.png)

Max projection:
![truncated_uint8_mean_enhanced](https://user-images.githubusercontent.com/28411193/174905724-59fb57ea-2b51-40d0-ae47-4c83b8e4e66c.png)

Correlation map:
![truncated_uint8_correlation_map](https://user-images.githubusercontent.com/28411193/174905827-8f63bec6-d960-4e6c-a85c-e394917f1d5a.png)

You can see the structured noise of the PMTs and scanning directions quite clearly here! These images could serve as a nice reference to show these artifacts as an example.

# Changing Exposure via `skimage.exposure.rescale_intensity`

I think I was using this incorrectly, but here's how I tried testing out this function from skimage that is supposed to rescale intensity values.

```
import h5py
import dask.array
import tifffile
from imagecodecs import tiff_decode
from skimage import exposure
from pathlib import Path


def tiff2hdf5(
    hdf5file, tifffiles, dataset_name='2p', chunksize=128, compression=None
):
    """Write images from TIFF files to chunked dataset in HDF5 file."""

    def imread(filename):
        # return first image in TIFF file as numpy array
        with open(filename, 'rb') as fh:
            data = fh.read()
        return tiff_decode(data)

    image_paths = [p for p in Path(tifffiles).glob("*Ch2*.tif")]
    sorted_paths = sorted(image_paths)

    store = tifffile.imread(image_paths, imread=imread, aszarr=True)
    images = dask.array.from_zarr(store)

    # I think this could be coercing the data into uint8 instead of doing the
    # rescaling, but since the final solution used worked I didn't investigate
    # it any further.
    transformed = images.map_blocks(exposure.rescale_intensity, dtype='uint8')

   transformed.to_hdf5('/scratch/file.hdf5', '/data', dtype='uint8')

```

Average Image:

![exposure_uint8_mean](https://user-images.githubusercontent.com/28411193/174906662-a06b2a66-ad4d-42e7-9c50-eeb122065dcb.png)

Average Enhanced:
![exposure_uint8_mean_enhanced](https://user-images.githubusercontent.com/28411193/174906674-65cc9cd7-9d0b-41d4-9a05-9ae300ba8ef6.png)

The enhanced mean shows the resonant scanning and PMT ripple noise extremely well here!

Very good example of it.

Correlation Map:
![exposure_uint8_correlation_map](https://user-images.githubusercontent.com/28411193/174906728-f430ac74-3278-4db0-b010-d09bdf4a6866.png)

Max Projection:

![exposure_uint8_max_projection](https://user-images.githubusercontent.com/28411193/174906782-87ae5708-9cc2-4cf9-ab6c-e7f160dc91ad.png)

# Doing it right: `img_as_ubyte`

skimage has a function for rescaling images to 8bit representations called `img_as_ubyte`. Mapping this to dask arrays makes it quite easy to translate the data into uint8 representation!

```
import h5py
import dask.array
import tifffile
from imagecodecs import tiff_decode
from skimage import img_as_ubyte
from pathlib import Path


def tiff2hdf5(
    hdf5file, tifffiles, dataset_name='2p', chunksize=128, compression=None
):
    """Write images from TIFF files to chunked dataset in HDF5 file."""

    def imread(filename):
        # return first image in TIFF file as numpy array
        with open(filename, 'rb') as fh:
            data = fh.read()
        return tiff_decode(data)

    image_paths = [p for p in Path(tifffiles).glob("*Ch2*.tif")]
    sorted_paths = sorted(image_paths)

    store = tifffile.imread(image_paths, imread=imread, aszarr=True)
    images = dask.array.from_zarr(store)

    # I think this could be coercing the data into uint8 instead of doing the
    # rescaling, but since the final solution used worked I didn't investigate
    # it any further.
    transformed = images.map_blocks(img_as_ubyte)

   transformed.to_hdf5('/scratch/file.hdf5', '/data', dtype='uint8')
```

This outputs data that looks just like the original but rescaled!

Average Image:
![correct_uint8_mean](https://user-images.githubusercontent.com/28411193/174907052-43a25b23-4897-4dd7-8321-09a392c5f2ab.png)

Average Enhanced:
![enhanced_correct_uint8](https://user-images.githubusercontent.com/28411193/174907098-3833e66d-cfce-48c7-8f50-5a40f7275628.png)

Correlation Map:
![correct_correlation_map_uint8](https://user-images.githubusercontent.com/28411193/174907125-8e47726a-e85e-4baf-beff-5ff997c588f4.png)

Max Projection:
![correct_max_projection_uint8](https://user-images.githubusercontent.com/28411193/174907136-1e7dd7f6-fa9f-4b99-b2a8-4b55e21c9f15.png)

**Comment by jmdelahanty** [comment]
New [Twitter](https://twitter.com/science_is_hard/status/1577863425677369344?s=20&t=GYHCI8YNGTeGKItFLd3AVQ) thread about lots of stuff that should be documented somewhere here on the repository/on a wiki somewhere ideally. Will update somewhere at some point...
